File: tests/c/src/T011_HeartbeatGenerator.c
#include "TestCase.h"


void verifyHeartbeatMessage(GMSEC_Config config, GMSEC_U16 expectedPubRate);


void test_heartbeatGeneratorCreate()
{
	GMSEC_INFO("Test heartbeatGeneratorCreate()");

	GMSEC_Status status = statusCreate();

	//o Nominal test
	{
		GMSEC_Config             config = configCreateCopy(testGetConfig(), NULL);
		GMSEC_HeartbeatGenerator hbgen  = heartbeatGeneratorCreate(config, (GMSEC_U16) 5, status);

		testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);
		testRequireBool("Heartbeat Generator handle is NULL", hbgen != NULL);
		testCheckBool("Heartbeat Generator should not be running", heartbeatGeneratorIsRunning(hbgen, NULL) == GMSEC_FALSE);

		heartbeatGeneratorDestroy(&hbgen);
		configDestroy(&config);
	}

	//o Off-nominal test(s)
	{
		// NULL config
		GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreate(NULL, (GMSEC_U16) 5, status);

		testRequireBool(statusGet(status), statusHasError(status) == GMSEC_TRUE);
		testRequireBool("Heartbeat Generator handle is NULL", hbgen == NULL);
	}

	statusDestroy(&status);
}


void test_heartbeatGeneratorCreateWithFields()
{
	GMSEC_INFO("Test heartbeatGeneratorCreateWithFields()");

	GMSEC_Status status = statusCreate();
	GMSEC_Field  fields[8];

	fields[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	fields[1] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST-ID", GMSEC_TRUE, NULL);
	fields[2] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[3] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	fields[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	fields[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	fields[7] = stringFieldCreate("COMPONENT", "T012", GMSEC_TRUE, NULL);

	//o Nominal test (fields provided)
	{
		GMSEC_Config             config = configCreateCopy(testGetConfig(), NULL);
		GMSEC_HeartbeatGenerator hbgen  = heartbeatGeneratorCreateWithFields(config, (GMSEC_U16) 5, fields, sizeof(fields)/sizeof(GMSEC_Field), status);

		testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);
		testRequireBool("Heartbeat Generator handle is NULL", hbgen != NULL);
		testCheckBool("Heartbeat Generator should not be running", heartbeatGeneratorIsRunning(hbgen, NULL) == GMSEC_FALSE);

		heartbeatGeneratorDestroy(&hbgen);
		configDestroy(&config);
	}

	//o Off-nominal test(s)
	{
		// NULL config
		GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(NULL, (GMSEC_U16) 5, fields, sizeof(fields)/sizeof(GMSEC_Field), status);

		testRequireBool(statusGet(status), statusHasError(status) == GMSEC_TRUE);
		testRequireBool("Heartbeat Generator handle is NULL", hbgen == NULL);
	}

	{
		// No fields provided
		GMSEC_Config             config = configCreateCopy(testGetConfig(), NULL);
		GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(config, (GMSEC_U16) 5, NULL, sizeof(fields)/sizeof(GMSEC_Field), status);

		testRequireBool(statusGet(status), statusHasError(status) == GMSEC_TRUE);
		testRequireBool("Heartbeat Generator handle is NULL", hbgen == NULL);

		configDestroy(&config);
	}

	int i;
	for (i = 0; i < sizeof(fields)/sizeof(GMSEC_Field); ++i)
	{
		fieldDestroy(&fields[i]);
	}
	statusDestroy(&status);
}


void test_heartbeatGeneratorDestroy()
{
	GMSEC_INFO("Test heartbeatGeneratorDestroy()");

	GMSEC_Config config = configCreateCopy(testGetConfig(), NULL);
	GMSEC_Field  fields[8];

	fields[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	fields[1] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST-ID", GMSEC_TRUE, NULL);
	fields[2] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[3] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	fields[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	fields[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	fields[7] = stringFieldCreate("COMPONENT", "T012", GMSEC_TRUE, NULL);

	GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(config, (GMSEC_U16) 5, fields, sizeof(fields)/sizeof(GMSEC_Field), NULL);

	testRequireBool("Heartbeat Generator handle is NULL", hbgen != NULL);

	heartbeatGeneratorDestroy(&hbgen);

	testRequireBool("Heartbeat Generator handle is not NULL", hbgen == NULL);

	int i;
	for (i = 0; i < sizeof(fields)/sizeof(GMSEC_Field); ++i)
	{
		fieldDestroy(&fields[i]);
	}
	configDestroy(&config);
}


void test_heartbeatGeneratorStart()
{
	GMSEC_INFO("Test heartbeatGeneratorStart()");

	GMSEC_Status status  = statusCreate();
	GMSEC_Config config  = testGetConfig();
	GMSEC_U16    pubRate = (GMSEC_U16) 1;
	GMSEC_Field  fields[8];

	fields[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	fields[1] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST-ID", GMSEC_TRUE, NULL);
	fields[2] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[3] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	fields[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	fields[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	fields[7] = stringFieldCreate("COMPONENT", "T012", GMSEC_TRUE, NULL);

	GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(config, pubRate, fields, sizeof(fields)/sizeof(GMSEC_Field), NULL);

	/* To ensure this heartbeat message is unique, we set the COMPONENT field */
	char tmp[OUT_STR_LEN];
	GMSEC_Field component = stringFieldCreate("COMPONENT", testGetUniqueComponent(tmp), GMSEC_TRUE, NULL);
	heartbeatGeneratorSetField(hbgen, component, NULL);
	fieldDestroy(&component);

	heartbeatGeneratorStart(hbgen, status);
	testCheckBool("Heartbeat Generator should be running", heartbeatGeneratorIsRunning(hbgen, NULL) == GMSEC_TRUE);

	verifyHeartbeatMessage(config, pubRate);

	// Off-nominal
	heartbeatGeneratorStart(NULL, status);
	testCheckBool("Expected an error", statusHasError(status) == GMSEC_TRUE);

	int i;
	for (i = 0; i < sizeof(fields)/sizeof(GMSEC_Field); ++i)
	{
		fieldDestroy(&fields[i]);
	}
	heartbeatGeneratorDestroy(&hbgen);
	statusDestroy(&status);
}


void test_heartbeatGeneratorStop()
{
	GMSEC_INFO("Test heartbeatGeneratorStop()");

	GMSEC_Status status  = statusCreate();
	GMSEC_Config config  = testGetConfig();
	GMSEC_U16    pubRate = (GMSEC_U16) 1;
	GMSEC_Field  fields[8];

	fields[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	fields[1] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST-ID", GMSEC_TRUE, NULL);
	fields[2] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[3] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	fields[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	fields[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	fields[7] = stringFieldCreate("COMPONENT", "T012", GMSEC_TRUE, NULL);

	GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(config, pubRate, fields, sizeof(fields)/sizeof(GMSEC_Field), NULL);

	/* To ensure this heartbeat message is unique, we set the COMPONENT field */
	char tmp[OUT_STR_LEN];
	GMSEC_Field component = stringFieldCreate("COMPONENT", testGetUniqueComponent(tmp), GMSEC_TRUE, NULL);
	heartbeatGeneratorSetField(hbgen, component, NULL);
	fieldDestroy(&component);

	heartbeatGeneratorStart(hbgen, NULL);
	testCheckBool("Heartbeat Generator should be running", heartbeatGeneratorIsRunning(hbgen, NULL) == GMSEC_TRUE);

	// Nominal test
	heartbeatGeneratorStop(hbgen, NULL);
	testCheckBool("Heartbeat Generator should not be running", heartbeatGeneratorIsRunning(hbgen, NULL) == GMSEC_FALSE);

	/* Allow time for the HB-gen thread to really stop */
	timeUtilMillisleep(2000);

	// Off-nominal test
	heartbeatGeneratorStop(NULL, status);
	testCheckBool("Expected an error", statusHasError(status) == GMSEC_TRUE);

	int i;
	for (i = 0; i < sizeof(fields)/sizeof(GMSEC_Field); ++i)
	{
		fieldDestroy(&fields[i]);
	}
	heartbeatGeneratorDestroy(&hbgen);
	statusDestroy(&status);
}


void test_heartbeatGeneratorIsRunning()
{
	GMSEC_INFO("Test heartbeatGeneratorIsRunning()");

	GMSEC_Status status = statusCreate();

	// Off-nominal tests (nominal are tested elsewhere)
	{
		GMSEC_BOOL isRunning = heartbeatGeneratorIsRunning(NULL, status);
		testCheckBool("Expected an error", statusHasError(status) == GMSEC_TRUE);
		testCheckBool("Expected isRunning to be false", isRunning == GMSEC_FALSE);
	}

	statusDestroy(&status);
}


void test_heartbeatGeneratorChangePublishRate()
{
	GMSEC_INFO("Test heartbeatGeneratorChangePublishRate()");

	GMSEC_Status status  = statusCreate();
	GMSEC_Config config  = testGetConfig();
	GMSEC_U16    pubRate = (GMSEC_U16) 1;
	GMSEC_Field  fields[8];

	fields[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	fields[1] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST-ID", GMSEC_TRUE, NULL);
	fields[2] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[3] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	fields[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	fields[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	fields[7] = stringFieldCreate("COMPONENT", "T012", GMSEC_TRUE, NULL);

	GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(config, pubRate, fields, sizeof(fields)/sizeof(GMSEC_Field), status);

	testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);

	/* To ensure this heartbeat message is unique, we set the COMPONENT field */
	char tmp[OUT_STR_LEN];
	GMSEC_Field component = stringFieldCreate("COMPONENT", testGetUniqueComponent(tmp), GMSEC_TRUE, NULL);
	heartbeatGeneratorSetField(hbgen, component, NULL);
	fieldDestroy(&component);

	heartbeatGeneratorStart(hbgen, NULL);

	// Nominal test
	heartbeatGeneratorChangePublishRate(hbgen, pubRate * 2, status);
	testCheckBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);

	verifyHeartbeatMessage(config, pubRate * 2);

	// Off-nominal test
	heartbeatGeneratorChangePublishRate(NULL, pubRate * 2, status);
	testCheckBool("Expected an error", statusHasError(status) == GMSEC_TRUE);

	int i;
	for (i = 0; i < sizeof(fields)/sizeof(GMSEC_Field); ++i)
	{
		fieldDestroy(&fields[i]);
	}
	heartbeatGeneratorDestroy(&hbgen);
	statusDestroy(&status);
}


void test_heartbeatGeneratorSetField()
{
	GMSEC_INFO("Test heartbeatGeneratorSetField()");

	GMSEC_Status status  = statusCreate();
	GMSEC_Config config  = testGetConfig();
	GMSEC_U16    pubRate = (GMSEC_U16) 1;
	GMSEC_Field  fields[8];

	fields[0] = stringFieldCreate("MISSION-ID", "MY-MISSION", GMSEC_TRUE, NULL);
	fields[1] = stringFieldCreate("CONSTELLATION-ID", "MY-CNST-ID", GMSEC_TRUE, NULL);
	fields[2] = stringFieldCreate("SAT-ID-PHYSICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[3] = stringFieldCreate("SAT-ID-LOGICAL", "MY-SAT-ID", GMSEC_TRUE, NULL);
	fields[4] = stringFieldCreate("FACILITY", "MY-FACILITY", GMSEC_TRUE, NULL);
	fields[5] = stringFieldCreate("DOMAIN1", "MY-DOMAIN-1", GMSEC_TRUE, NULL);
	fields[6] = stringFieldCreate("DOMAIN2", "MY-DOMAIN-2", GMSEC_TRUE, NULL);
	fields[7] = stringFieldCreate("COMPONENT", "T012", GMSEC_TRUE, NULL);

	GMSEC_HeartbeatGenerator hbgen = heartbeatGeneratorCreateWithFields(config, pubRate, fields, sizeof(fields)/sizeof(GMSEC_Field), status);

	testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);

	/* To ensure this heartbeat message is unique, we set the COMPONENT field */
	char tmp[OUT_STR_LEN];
	GMSEC_Field component = stringFieldCreate("COMPONENT", testGetUniqueComponent(tmp), GMSEC_TRUE, NULL);
	heartbeatGeneratorSetField(hbgen, component, NULL);
	fieldDestroy(&component);

	heartbeatGeneratorStart(hbgen, NULL);

	// Nominal tests
	{
		GMSEC_Field newPubRate = u16FieldCreate("PUB-RATE", pubRate * 2, GMSEC_FALSE, NULL);
		heartbeatGeneratorSetField(hbgen, newPubRate, status);
		testCheckBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);
		verifyHeartbeatMessage(config, pubRate * 2);
		fieldDestroy(&newPubRate);

		newPubRate = i32FieldCreate("PUB-RATE", (GMSEC_I32) pubRate * 3, GMSEC_FALSE, NULL);
		heartbeatGeneratorSetField(hbgen, newPubRate, status);
		testCheckBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);
		verifyHeartbeatMessage(config, pubRate * 3);
		fieldDestroy(&newPubRate);

		newPubRate = stringFieldCreate("PUB-RATE", "2", GMSEC_FALSE, NULL);
		heartbeatGeneratorSetField(hbgen, newPubRate, status);
		testCheckBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);
		verifyHeartbeatMessage(config, pubRate * 2);
		fieldDestroy(&newPubRate);
	}

	// Off-nominal tests
	{
		// HeartbeatGenerator is NULL
		GMSEC_Field newPubRate = u16FieldCreate("PUB-RATE", pubRate * 2, GMSEC_FALSE, NULL);
		heartbeatGeneratorSetField(NULL, newPubRate, status);
		testCheckBool("Expected an error", statusHasError(status) == GMSEC_TRUE);
		fieldDestroy(&newPubRate);

		heartbeatGeneratorSetField(hbgen, NULL, status);
		testCheckBool("Expected an error", statusHasError(status) == GMSEC_TRUE);
	}

	{
		// PUB-RATE is negative
		GMSEC_Field newPubRate = i16FieldCreate("PUB-RATE", -1, GMSEC_FALSE, NULL);
		heartbeatGeneratorSetField(hbgen, newPubRate, status);
		testCheckBool("Expected an error", statusHasError(status) == GMSEC_TRUE);
		fieldDestroy(&newPubRate);
	}

	int i;
	for (i = 0; i < sizeof(fields)/sizeof(GMSEC_Field); ++i)
	{
		fieldDestroy(&fields[i]);
	}
	heartbeatGeneratorDestroy(&hbgen);
	statusDestroy(&status);
}


int test_HeartbeatGenerator()
{
	testSetDescription("Test HeartbeatGenerator");

	test_heartbeatGeneratorCreate();
	test_heartbeatGeneratorCreateWithFields();
	test_heartbeatGeneratorDestroy();
	test_heartbeatGeneratorStart();
	test_heartbeatGeneratorStop();
	test_heartbeatGeneratorIsRunning();
	test_heartbeatGeneratorChangePublishRate();
	test_heartbeatGeneratorSetField();

	return 0;
}
TEST_DRIVER(test_HeartbeatGenerator)


void verifyHeartbeatMessage(GMSEC_Config config, GMSEC_U16 expectedPubRate)
{
	GMSEC_Status     status = statusCreate();
	GMSEC_Connection conn   = connectionCreate(config, status);
	char             topic[OUT_STR_LEN];
	char             tmp[OUT_STR_LEN];
	double           t1 = 0, t2;
	int              i;

	// Ensure we have a connection object
	testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);

	printToString(topic, "C2MS.*.*.*.*.*.MSG.HB.%s.+", testGetUniqueComponent(tmp));

	// Ensure we can connect to the bus
	connectionConnect(conn, status);
	testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);

	// Subscribe to receive messages
	connectionSubscribe(conn, topic, status);
	testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);

	for (i = 0; i < 7; ++i)
	{
		// Ensure we are still connected
		GMSEC_Message msg = connectionReceive(conn, 5000, status);
		testRequireBool(statusGet(status), statusHasError(status) == GMSEC_FALSE);

		// Ignore the first few incoming messages (if any)
		if (i < 3)
		{
			if (msg != NULL) messageDestroy(&msg);
			continue;
		}

		if (msg != NULL)
		{
			if (t1 == 0)
			{
				t1 = timeUtilGetCurrentTime_s(NULL);
			}
			else
			{
				t2 = timeUtilGetCurrentTime_s(NULL);

				double delta = t2 - t1;
				if (delta < expectedPubRate)
				{
					delta = ((t2 - t1) * 10.0 + 0.5) / 10.0;
				}

				GMSEC_INFO("Expected rate is %u, delta is: %g", expectedPubRate, delta);

				testCheckBool("Unexpected publish rate", expectedPubRate == (GMSEC_U16)delta);

				t1 = t2;
			}

			testCheckBool("Unexpected MESSAGE-TYPE", strcompare(messageGetStringValue(msg, "MESSAGE-TYPE", NULL), "MSG") == 0);
			testCheckBool("Unexpected MESSAGE-SUBTYPE", strcompare(messageGetStringValue(msg, "MESSAGE-SUBTYPE", NULL), "HB") == 0);
			testCheckBool("Unexpected PUB-RATE", messageGetU32Value(msg, "PUB-RATE", NULL) == (GMSEC_U32) expectedPubRate);

			messageDestroy(&msg);
		}
		else
		{
			testCheckBool("Failed to receive heartbeat message", GMSEC_FALSE);
		}
	}

	connectionDestroy(&conn);
	statusDestroy(&status);
}

File: tests/perl/src/T011_HeartbeatGenerator.pl
#!/usr/bin/env perl

use strict;
use Test;
use libgmsec_perl;


sub Test_HeartbeatGenerator
{
	my ($test) = @_;

	$test->setDescription("Test HeartbeatGenerator");

	test_create($test);
	test_start($test);
	test_stop($test);
	test_changePublishRate($test);
	test_setField($test);
}


sub test_create
{
	libgmsec_perl::logInfo("Test create()");

	my ($test) = @_;

	my $emptyFieldArray = libgmsec_perl::FieldArray->new();
	my $standardFields  = $test->getStandardFields();

	# Nominal test
	my $hbgen1 = libgmsec_perl::HeartbeatGenerator::create($test->getConfig(), 5, $emptyFieldArray);
	my $hbgen2 = libgmsec_perl::HeartbeatGenerator::create($test->getConfig(), 5, $standardFields);
	my $hbgen3 = libgmsec_perl::HeartbeatGenerator::create($test->getConfig(), 5);

	$test->check("HeartbeatGenerator should not be running", $hbgen1->isRunning() == 0);
	$test->check("HeartbeatGenerator should not be running", $hbgen2->isRunning() == 0);
	$test->check("HeartbeatGenerator should not be running", $hbgen3->isRunning() == 0);

	libgmsec_perl::HeartbeatGenerator::destroy($hbgen1);
	libgmsec_perl::HeartbeatGenerator::destroy($hbgen2);
	libgmsec_perl::HeartbeatGenerator::destroy($hbgen3);

    # Off-nominal test(s)
    eval
    {
        my $config = libgmsec_perl::Config->new($test->getConfig());
        $config->addValue("mw-id", "bogus-mw");
    	my $hbgen = libgmsec_perl::HeartbeatGenerator::create($config, 5, $emptyFieldArray);
	};
	if (isa($@, 'libgmsec_perl::GmsecException'))
	{
		my $error = $@;
		$test->check($error->what(), index($error->what(), "Unable to load") != -1);
	}
}


sub test_start
{
	libgmsec_perl::logInfo("Test start()");

	my ($test) = @_;

	my $config = libgmsec_perl::Config->new($test->getConfig());

	my $hbgen = libgmsec_perl::HeartbeatGenerator::create($config, 1, $test->getStandardFields());

	# To ensure this heartbeat message is unique, we set the COMPONENT field
	$hbgen->setField( libgmsec_perl::StringField->new("COMPONENT", $test->getUniqueComponent()) );

	$hbgen->start();

	$test->check("HeartbeatGenerator should be running", $hbgen->isRunning());

	verifyHeartbeatMessage($test, $config, 1);

	$hbgen->stop();

	libgmsec_perl::HeartbeatGenerator::destroy($hbgen);

	# Off-nominal tests
	libgmsec_perl::logInfo("Off-nominal cases...");
	$config->addValue("gmsec-msg-content-validate", "true");

	my $hbgen2 = libgmsec_perl::HeartbeatGenerator::create($config, 1, $test->getStandardFields());

	# Add bogus field using a Field
	eval
	{
		$hbgen2->setField( libgmsec_perl::U16Field->new("BOGUS-FIELD", 2) );
        $hbgen2->start();
		$test->check("An expection was expected", 0);
	};
	if (isa($@, 'libgmsec_perl::GmsecException'))
	{
		my $error = $@;
		$test->check($error->what(), index($error->what(), "Message Validation Failed") != -1);
	}

	# Add bogus field using a long value
	eval
	{
		$hbgen2->setField("BOGUS-FIELD", 2);
        $hbgen2->start();
		$test->check("An expection was expected", 0);
	};
	if (isa($@, 'libgmsec_perl::GmsecException'))
	{
		my $error = $@;
		$test->check($error->what(), index($error->what(), "Message Validation Failed") != -1);
	}

	# Add bogus field using a double value
	eval
	{
		$hbgen2->setField("BOGUS-FIELD", 2.0);
        $hbgen2->start();
		$test->check("An expection was expected", 0);
	};
	if (isa($@, 'libgmsec_perl::GmsecException'))
	{
		my $error = $@;
		$test->check($error->what(), index($error->what(), "Message Validation Failed") != -1);
	}

	# Add bogus field using a string value
	eval
	{
		$hbgen2->setField("BOGUS-FIELD", "2");
        $hbgen2->start();
		$test->check("An expection was expected", 0);
	};
	if (isa($@, 'libgmsec_perl::GmsecException'))
	{
		my $error = $@;
		$test->check($error->what(), index($error->what(), "Message Validation Failed") != -1);
	}

	libgmsec_perl::HeartbeatGenerator::destroy($hbgen2);
}


sub test_stop
{
	libgmsec_perl::logInfo("Test stop()");

	my ($test) = @_;

	my $config = $test->getConfig();

	my $hbgen = libgmsec_perl::HeartbeatGenerator::create($config, 1, $test->getStandardFields());

	# To ensure this heartbeat message is unique, we set the COMPONENT field
	$hbgen->setField( libgmsec_perl::StringField->new("COMPONENT", $test->getUniqueComponent()) );

	$hbgen->start();

	$test->check("HeartbeatGenerator should be running", $hbgen->isRunning());

	$hbgen->stop();

	$test->check("HeartbeatGenerator should not be running", $hbgen->isRunning() == 0);

	libgmsec_perl::TimeUtil::millisleep(2000);

	my $conn = libgmsec_perl::Connection::create($config);
	$conn->connect();
	$conn->subscribe("C2MS.*.*.*.*.*.MSG.HB." . $test->getUniqueComponent());

	for (my $i = 0; $i < 3; $i++)
	{
		my $hbmsg = $conn->receive(5000);

		if ($hbmsg != undef)
		{
			$test->check("Unexpectedly received a Heartbeat Message", 0);
			libgmsec_perl::Message::destroy($hbmsg);
		}
	}

	$conn->disconnect();

	libgmsec_perl::Connection::destroy($conn);

	libgmsec_perl::HeartbeatGenerator::destroy($hbgen);
}


sub test_changePublishRate
{
	libgmsec_perl::logInfo("Test changePublishRate()");

	my ($test) = @_;

	my $config = $test->getConfig();

	# Initialize set rate to 1 (second)
	my $hbgen = libgmsec_perl::HeartbeatGenerator::create($config, 1, $test->getStandardFields());

	# To ensure this heartbeat message is unique, we set the COMPONENT field
	$hbgen->setField( libgmsec_perl::StringField->new("COMPONENT", $test->getUniqueComponent()) );

	$hbgen->start();

	# Change rate to 2 (seconds)
	$hbgen->changePublishRate(2);
	verifyHeartbeatMessage($test, $config, 2);

	my $conn = libgmsec_perl::Connection::create($config);
	$conn->connect();
	$conn->subscribe("C2MS.*.*.*.*.*.MSG.HB." . $test->getUniqueComponent());

	libgmsec_perl::TimeUtil::millisleep(1000);

	# Change rate to 0 (seconds); expect only one additional heartbeat message to be published
	$hbgen->changePublishRate(0);

	libgmsec_perl::TimeUtil::millisleep(1000);

	my $lastMessage = 0;

	for (my $i = 0; $i < 3 and $lastMessage == 0; $i++)
	{
		my $hbmsg = $conn->receive(5000);

		if ($hbmsg == undef) 
		{
			$test->check("Failed to receive Heartbeat Message", 0);
		}
		elsif ($hbmsg->getIntegerValue("PUB-RATE") != 0)
		{
			libgmsec_perl::Message::destroy($hbmsg);
		}
		else
		{
			$lastMessage = 1;
			libgmsec_perl::Message::destroy($hbmsg);
		}
	}

	$test->check("Failed to receive message with PUB-RATE of 0", $lastMessage == 1);

	$conn->disconnect();

	libgmsec_perl::Connection::destroy($conn);

	$hbgen->stop();

	libgmsec_perl::HeartbeatGenerator::destroy($hbgen);
}


sub test_setField
{
	my ($test) = @_;

	my $config = libgmsec_perl::Config->new( $test->getConfig() );     # make a copy

	my $pubRate = 1;

	# Nominal tests
	my $hbgen = libgmsec_perl::HeartbeatGenerator::create($config, $pubRate, $test->getStandardFields());

	# To ensure this heartbeat message is unique, we set the COMPONENT field
	$hbgen->setField( libgmsec_perl::StringField->new("COMPONENT", $test->getUniqueComponent()) );

	$hbgen->start();

	libgmsec_perl::TimeUtil::millisleep(2000);

	libgmsec_perl::logInfo("Test setField() with a Field");
	$hbgen->setField( libgmsec_perl::U16Field->new("PUB-RATE", 2) );
	verifyHeartbeatMessage($test, $config, 2);

	libgmsec_perl::logInfo("Test setField() with a Field Name and long value");
	$hbgen->setField("PUB-RATE", 3);
	verifyHeartbeatMessage($test, $config, 3);

	libgmsec_perl::logInfo("Test setField() with a Field Name and double value");
	$hbgen->setField("PUB-RATE", 2.0);
	verifyHeartbeatMessage($test, $config, 2);

	libgmsec_perl::logInfo("Test setField() with a Field Name and string value");
	$hbgen->setField("PUB-RATE", "3");
	verifyHeartbeatMessage($test, $config, 3);

	$hbgen->stop();

	libgmsec_perl::logInfo("Cleanup...");
	libgmsec_perl::HeartbeatGenerator::destroy($hbgen);
}


sub verifyHeartbeatMessage
{
	my ($test, $config, $expectedPubRate) = @_;

	my $conn = libgmsec_perl::Connection::create($config);
	$conn->connect();
	$conn->subscribe("C2MS.*.*.*.*.*.MSG.HB." . $test->getUniqueComponent() . ".+");

	my $t1 = 0;
	my $t2 = 0;

	for (my $i = 0; $i < 7; $i++)
	{
		my $hbmsg = $conn->receive(5000);

		# ignore the first few incoming messages (if any)
		if ($i < 3)
		{
			if (defined $hbmsg)
			{
				libgmsec_perl::Message::destroy($hbmsg);
			}
			next;
		}

		if ($hbmsg == undef)
		{
			$test->check("Failed to received Heartbeat Message", 0);
		}
		else
		{
			if ($t1 == 0)
			{
				$t1 = libgmsec_perl::TimeUtil::getCurrentTime_s();
			}
			else
			{
				$t2 = libgmsec_perl::TimeUtil::getCurrentTime_s();

                my $delta = $t2 - $t1;
                if ($delta < $expectedPubRate)
                {
                    $delta = (($t2 - $t1) * 10.0 + 0.5) / 10.0;
                }

				libgmsec_perl::logInfo("Expected rate is: $expectedPubRate, delta is: $delta");

				$test->check("Unexpected publish rate", int($expectedPubRate) == int($delta));

				$t1 = $t2;

				$test->check("Unexpected MESSAGE-TYPE", $hbmsg->getStringValue("MESSAGE-TYPE") eq "MSG");
				$test->check("Unexpected MESSAGE-SUBTYPE", $hbmsg->getStringValue("MESSAGE-SUBTYPE") eq "HB");
				$test->check("Unexpected PUB-RATE", $hbmsg->getIntegerValue("PUB-RATE") == $expectedPubRate);

				libgmsec_perl::Message::destroy($hbmsg);
			}
		}
	}

	$conn->disconnect();

	libgmsec_perl::Connection::destroy($conn);
}


Test::driver(\&Test_HeartbeatGenerator);
